pydocstringformatter/run.py

```diff
@@ -98,26 +84,148 @@ def format_file(self, filename: Path) -> bool:
                         file=sys.stderr,
                     )
                 with open(filename, "w", encoding="utf-8", newline=newlines) as file:
-                    file.write(tokenize.untokenize(changed_tokens))
+                    file.write(tokenize.untokenize(formatted_tokens))
                     _utils.print_to_console(
                         f"Formatted {filename_str} 📖\n", self.config.quiet
                     )
             else:
                 sys.stdout.write(
                     _utils.generate_diff(
                         tokenize.untokenize(tokens),
-                        tokenize.untokenize(changed_tokens),
+                        tokenize.untokenize(formatted_tokens),
                         filename_str,
                     )
                 )
 
         return is_changed
 
-    def format_files(self, filepaths: list[Path]) -> bool:
-        """Format a list of files."""
+    def format_file_tokens(
+        self, tokens: list[tokenize.TokenInfo], filename: Path
+    ) -> tuple[list[tokenize.TokenInfo], bool]:
+        """Format a list of tokens.
+
+        tokens: List of tokens to format.
+        filename: Name of the file the tokens are from.
+
+        Returns:
+            A tuple containing [1] the formatted tokens in a list
+            and a boolean indicating if the tokens were changed.
+
+        Raises:
+            RuntimeError:
+                If the formatters are not able to get to a stable
+                result after 3 iterations.
+
+                It reports what formatters are still modifying the tokens.
+        """
+        formatted_tokens: list[tokenize.TokenInfo] = []
         is_changed = False
 
-        for file in filepaths:
-            is_changed = self.format_file(file) or is_changed
+        for index, tokeninfo in enumerate(tokens):
+            new_tokeninfo = tokeninfo
 
-        return is_changed
+            if _utils.is_docstring(new_tokeninfo, tokens[index - 1]):
+                new_tokeninfo, changers, stable = self.apply_formatters_twice(
+                    new_tokeninfo
+                )
+                is_changed = is_changed or bool(changers)
+
+                # Try again to get a stable result, skipping the formatters that
```

Comment:
I think this might be problematic.

Let's say we have:
```python
"""My summary
is two lines.
"""
```

This formatter that puts a period at the end of the summary won't do anything in this case.
However, if after the first run the summary splitter creates:

```python
"""My summary

is two lines.
"""
```
Then the period added should be called again.

So, I don't think we can skip formatters based on results in a previous round.

Reply:
I gave this a lot of thought and this is why every pass is "twice" (hence apply_formatters_twice).

For example if our formatters are ...
[AtoB, BtoC, CtoA]
for "B"
first pass: [B>C, C>A]
second pass: [A>B,B>C,C>A]

Therefore all the formatters will be flagged as changers in the "first double pass". So all of them will be used again in L145.

The tradeoff here is catching some extremely complicated multi-formatter conflict that would only arise after 3 full passes vs speed.

Reply:
Take:
[DtoB, AtoB, BtoC, CtoA]
for "D"
first pass: [D>B,B>C, C>A]
second pass: [A>B,B>C,C>A]

So the `DtoB` won't be called again in the third case. Imo, that time save is negligible in comparison to the complexity this adds. `pydocstringformatter` is (currently) quite fast anyways. Especially when there are no changes to write which will be the case for most projects after they have done their initial setup. I'd prefer just running all conflicts in all 3 runs and having that slight bit of potential delay.
It might also be hard to debug issues if not all 3 of the runs are the same. You'd have to figure out which formatters are turned on when to do a full analysis which might be annoying.

Reply:
alright I see your point.
Will remove that filter in the "second double pass"

Reply:
Done